File: Introduction to Robotics Course/Chapter7/Lesson5/Chapter7_Lesson5.cpp
#include <iostream>
#include <vector>
#include <random>
#include <cmath>

int main() {
    double fs = 100.0, Ts = 1.0/fs;
    int N = 200;
    std::vector<double> x(N), b(N), y(N), yq(N), t(N);

    // Random generators
    std::mt19937 gen(0);
    std::normal_distribution<double> noise(0.0, 0.05);
    std::normal_distribution<double> drift(0.0, 1e-4);

    // Quantizer parameters
    int B = 10;
    double ymin = -2.0, ymax = 2.0;
    double Delta = (ymax - ymin) / std::pow(2.0, B);

    // Simulate
    for(int k=0; k<N; ++k){
        t[k] = k*Ts;
        x[k] = std::sin(2*M_PI*3*t[k]);               // assume band-limited here
        if(k==0) b[k] = 0.0; else b[k] = b[k-1] + drift(gen);
        y[k] = x[k] + b[k] + noise(gen);

        // Quantize
        yq[k] = Delta * std::round(y[k]/Delta);
    }

    // Estimate bias from first 20 samples
    double b_hat = 0.0;
    for(int k=0; k<20; ++k) b_hat += yq[k];
    b_hat /= 20.0;

    std::cout << "Delta=" << Delta << ", b_hat=" << b_hat << std::endl;
    return 0;
}

File: Introduction to Robotics Course/Chapter7/Lesson5/Chapter7_Lesson5.py
import numpy as np

# Continuous-time "true" signal sampled very finely for simulation
fs_true = 5000.0
T_true = 1.0 / fs_true
t = np.arange(0, 2.0, T_true)
x = np.sin(2*np.pi*3*t) + 0.4*np.sin(2*np.pi*40*t)  # low + high freq content

# Sensor noise and drift models
sigma_n = 0.05
sigma_b = 1e-4
n = np.random.normal(0, sigma_n, size=t.size)

b = np.zeros_like(t)
for k in range(1, t.size):
    b[k] = b[k-1] + np.random.normal(0, sigma_b)

y_analog = x + n + b

# Sampling
fs = 100.0
Ts = 1.0 / fs
idx = (t / Ts).astype(int)
idx = np.unique(idx)
t_s = t[idx]
y_s = y_analog[idx]

# Quantization (B bits on range [-2,2])
B = 10
ymin, ymax = -2.0, 2.0
Delta = (ymax - ymin) / (2**B)
y_q = Delta * np.round(y_s / Delta)

print("Quantization step Delta:", Delta)
print("Measured variance (noise+quant):", np.var(y_q - np.sin(2*np.pi*3*t_s)))

# Simple bias estimation by averaging over a known "static" interval
# Here we pretend first 0.2s is static so x≈0 there.
static_mask = t_s < 0.2
b_hat = np.mean(y_q[static_mask])
y_comp = y_q - b_hat
print("Estimated bias:", b_hat)
